### ISSUE TITLE
Implement ExecutorService for multithreading

### ISSUE THREAD

**Issue by vibhcool** [loklak/loklak_server#1330]
- Issue type: Feature request

The task is to implement ExecutorService for threads to optimize the performance of multithreading in Loklak Server

**Comment by vibhcool** [comment]
working on this

**Comment by DengYiping** [comment]
Hello! 

I am interested in enhancing the multithreading in loklak, do you have any suggestion for it? My point of view is that we should use a new thread-memory model, such as actor and reactor model, to create a boundary for each subsystem. 

I'm quiet familiar with Akka, the leading actor model implementation on Java, and **moving to Akka, without a doubt, can solve this issue**, but this project is gonna be huge.

We should discuss it before taking any move on this issue. @vibhcool @jayvdb @pguth @Orbiter @pguth What do u guys think of this idea? 

BTW I have rewritten version of loklak(just the crawler part) that runs on Akka and Scala, you can have an idea of the code structure if you go into my repo to take a look.

**Comment by vibhcool** [comment]
At present, the main issue with Loklak Server is very high memory usage and low CPU consumption. Can Akka be used to solve this issue? 
the server consumes about 20% of memory and about 90% of memory which makes it vulnerable of `Error java.lang.OutOfMemoryError: GC overhead limit exceeded`

**Comment by vibhcool** [comment]
This issue will prevent us from adding higher memory application features. IMO, if CPU usage is increased, less memory will be consumed an instant. 

views?

**Comment by DengYiping** [comment]
I see, the threshold is **memory storage**, right? Well, I think we are not in a situation that Akka has to be introduced. (For not doubt, Akka will lower the overall memory use, but moving to Akka is a big project)
I have a better solution for this little problem.


Loklak is divided into two parts: 
`scraper, web service`  and   `Elasticsearch node` 
To low the memory usage, my suggestion is that we should move elasticsearch node out of the Java application, because a elasticsearch node (or node driver) consumes huge amount of memory.  


Elasticsearch should run as a service, and in the Java application, we use a elasticsearch Transport Client **(not node client)** to connect to the ES service, then we can do optimisation on elasticsearch(use a better configuration, caching, memory compression etc.).

If the memory usage is still high, we should check if the memory is leaked in loklak, or use Java off-heap storage.

**Comment by DengYiping** [comment]
@vibhcool @jayvdb @pguth @Orbiter @pguth Do you guys think that we should use a Elasticsearch transport client and move out the heavy elasticsearch to a service?

**Comment by vibhcool** [comment]
@DengYiping , One approach I came up was solving the issue by setting up a pattern or formula for configuration of the setting values of ElasticSearch parameters and various threads.
views about tuning the configurations?

IMHO, we can move out ElasticSearch at the main deployment, but as the software, It shall be part of the package as it is a part of a unit loklak server. Also Loklak isn't a very big software, It can be tuned for better resource management.

**Comment by DengYiping** [comment]
@vibhcool I agree with your first point, so we will not move out ElasticSearch then.

 If it is ElasticSearch that consumes huge amount of memory, the only thing we can do is tuning it. If other parts of loklak consumes the majority of the memory, better resource management can solve the problem. 

I think we should start to benchmark loklak and find out which part consumes most of the memory( my thought is the ElasticSearch part, but I hope it is other parts of loklak)
